<!DOCTYPE html>
<html lang="en-us">
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1">
    
    <title>Data Wrangling with R Basics - 2 | Ananke</title>
    <meta name="viewport" content="width=device-width,minimum-scale=1">
    <meta name="description" content="Working with genomic data with R We provided 5 files in the Data directory: - differntial_results.csv - gene_expression.tsv - gene_annotations.tsv
my_pathway_genes.txt - mutations.tsv These files are stored in either comma-delimited (csv) or tab-delimited (tsv) formats. Files can be read with the following functions available from default R libraries. These are built-in “functions” from R. All read files, but with slightly different default behavior.
# setwd(&quot;~/modules/data_wrangling&quot;) # First, load all the data: print(&#39;Loading data.">
    <meta name="generator" content="Hugo 0.101.0" />
    
    
    
    
      <meta name="robots" content="noindex, nofollow">
    

    
<link rel="stylesheet" href="/ananke/css/main.min.css" >



    
    
    
      

    

    
    
    <meta property="og:title" content="Data Wrangling with R Basics - 2" />
<meta property="og:description" content="Working with genomic data with R We provided 5 files in the Data directory: - differntial_results.csv - gene_expression.tsv - gene_annotations.tsv
my_pathway_genes.txt - mutations.tsv These files are stored in either comma-delimited (csv) or tab-delimited (tsv) formats. Files can be read with the following functions available from default R libraries. These are built-in “functions” from R. All read files, but with slightly different default behavior.
# setwd(&quot;~/modules/data_wrangling&quot;) # First, load all the data: print(&#39;Loading data." />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://yaoyu-e-wang.github.io/posts/datawranglingbasic-2/" /><meta property="article:section" content="posts" />
<meta property="article:published_time" content="2022-04-25T00:00:00+00:00" />
<meta property="article:modified_time" content="2022-04-25T00:00:00+00:00" /><meta property="og:site_name" content="Ananke" />

<meta itemprop="name" content="Data Wrangling with R Basics - 2">
<meta itemprop="description" content="Working with genomic data with R We provided 5 files in the Data directory: - differntial_results.csv - gene_expression.tsv - gene_annotations.tsv
my_pathway_genes.txt - mutations.tsv These files are stored in either comma-delimited (csv) or tab-delimited (tsv) formats. Files can be read with the following functions available from default R libraries. These are built-in “functions” from R. All read files, but with slightly different default behavior.
# setwd(&quot;~/modules/data_wrangling&quot;) # First, load all the data: print(&#39;Loading data."><meta itemprop="datePublished" content="2022-04-25T00:00:00+00:00" />
<meta itemprop="dateModified" content="2022-04-25T00:00:00+00:00" />
<meta itemprop="wordCount" content="2392">
<meta itemprop="keywords" content="R,Data Wrangling,tutorial," /><meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="Data Wrangling with R Basics - 2"/>
<meta name="twitter:description" content="Working with genomic data with R We provided 5 files in the Data directory: - differntial_results.csv - gene_expression.tsv - gene_annotations.tsv
my_pathway_genes.txt - mutations.tsv These files are stored in either comma-delimited (csv) or tab-delimited (tsv) formats. Files can be read with the following functions available from default R libraries. These are built-in “functions” from R. All read files, but with slightly different default behavior.
# setwd(&quot;~/modules/data_wrangling&quot;) # First, load all the data: print(&#39;Loading data."/>

	
  </head>

  <body class="ma0 avenir bg-near-white">

    
   
  

  <header>
    <div class="bg-black">
      <nav class="pv3 ph3 ph4-ns" role="navigation">
  <div class="flex-l justify-between items-center center">
    <a href="/" class="f3 fw2 hover-white no-underline white-90 dib">
      
        Ananke
      
    </a>
    <div class="flex-l items-center">
      

      
        <ul class="pl0 mr3">
          
          <li class="list f5 f4-ns fw4 dib pr3">
            <a class="hover-white no-underline white-90" href="/posts/" title="Posts page">
              Posts
            </a>
          </li>
          
        </ul>
      
      
<div class="ananke-socials">
  
    <a href="https://twitter.com/GoHugoIO" target="_blank" class="twitter ananke-social-link link-transition stackoverflow link dib z-999 pt3 pt0-l mr1" title="Twitter link" rel="noopener" aria-label="follow on Twitter——Opens in a new window">
      
        <span class="icon"><svg style="enable-background:new 0 0 67 67;" version="1.1" viewBox="0 0 67 67"  xml:space="preserve" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"><path d="M37.167,22.283c-2.619,0.953-4.274,3.411-4.086,6.101  l0.063,1.038l-1.048-0.127c-3.813-0.487-7.145-2.139-9.974-4.915l-1.383-1.377l-0.356,1.017c-0.754,2.267-0.272,4.661,1.299,6.271  c0.838,0.89,0.649,1.017-0.796,0.487c-0.503-0.169-0.943-0.296-0.985-0.233c-0.146,0.149,0.356,2.076,0.754,2.839  c0.545,1.06,1.655,2.097,2.871,2.712l1.027,0.487l-1.215,0.021c-1.173,0-1.215,0.021-1.089,0.467  c0.419,1.377,2.074,2.839,3.918,3.475l1.299,0.444l-1.131,0.678c-1.676,0.976-3.646,1.526-5.616,1.568  C19.775,43.256,19,43.341,19,43.405c0,0.211,2.557,1.397,4.044,1.864c4.463,1.377,9.765,0.783,13.746-1.568  c2.829-1.673,5.657-5,6.978-8.221c0.713-1.716,1.425-4.851,1.425-6.354c0-0.975,0.063-1.102,1.236-2.267  c0.692-0.678,1.341-1.419,1.467-1.631c0.21-0.403,0.188-0.403-0.88-0.043c-1.781,0.636-2.033,0.551-1.152-0.402  c0.649-0.678,1.425-1.907,1.425-2.267c0-0.063-0.314,0.042-0.671,0.233c-0.377,0.212-1.215,0.53-1.844,0.72l-1.131,0.361l-1.027-0.7  c-0.566-0.381-1.361-0.805-1.781-0.932C39.766,21.902,38.131,21.944,37.167,22.283z M33,64C16.432,64,3,50.569,3,34S16.432,4,33,4  s30,13.431,30,30S49.568,64,33,64z" style="fill-rule:evenodd;clip-rule:evenodd;"/></svg>
</span>
      
<span class="new-window"><svg  height="8px"  style="enable-background:new 0 0 1000 1000;" version="1.1" viewBox="0 0 1000 1000"  xml:space="preserve" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" >
<path d="M598 128h298v298h-86v-152l-418 418-60-60 418-418h-152v-86zM810 810v-298h86v298c0 46-40 86-86 86h-596c-48 0-86-40-86-86v-596c0-46 38-86 86-86h298v86h-298v596h596z" style="fill-rule:evenodd;clip-rule:evenodd;"/>
</svg>
</span></a>
  
</div>

    </div>
  </div>
</nav>

    </div>
  </header>



    <main class="pb7" role="main">
      
  
  <article class="flex-l flex-wrap justify-between mw8 center ph3">
    <header class="mt4 w-100">
      <aside class="instapaper_ignoref b helvetica tracked">
          
        POSTS
      </aside>
      










  <div id="sharing" class="mt3 ananke-socials">
    
      
      <a href="https://twitter.com/share?url=https://yaoyu-e-wang.github.io/posts/datawranglingbasic-2/&amp;text=Data%20Wrangling%20with%20R%20Basics%20-%202" class="ananke-social-link twitter no-underline" aria-label="share on Twitter">
        
        <span class="icon"> <svg style="enable-background:new 0 0 67 67;" version="1.1" viewBox="0 0 67 67"  xml:space="preserve" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"><path d="M37.167,22.283c-2.619,0.953-4.274,3.411-4.086,6.101  l0.063,1.038l-1.048-0.127c-3.813-0.487-7.145-2.139-9.974-4.915l-1.383-1.377l-0.356,1.017c-0.754,2.267-0.272,4.661,1.299,6.271  c0.838,0.89,0.649,1.017-0.796,0.487c-0.503-0.169-0.943-0.296-0.985-0.233c-0.146,0.149,0.356,2.076,0.754,2.839  c0.545,1.06,1.655,2.097,2.871,2.712l1.027,0.487l-1.215,0.021c-1.173,0-1.215,0.021-1.089,0.467  c0.419,1.377,2.074,2.839,3.918,3.475l1.299,0.444l-1.131,0.678c-1.676,0.976-3.646,1.526-5.616,1.568  C19.775,43.256,19,43.341,19,43.405c0,0.211,2.557,1.397,4.044,1.864c4.463,1.377,9.765,0.783,13.746-1.568  c2.829-1.673,5.657-5,6.978-8.221c0.713-1.716,1.425-4.851,1.425-6.354c0-0.975,0.063-1.102,1.236-2.267  c0.692-0.678,1.341-1.419,1.467-1.631c0.21-0.403,0.188-0.403-0.88-0.043c-1.781,0.636-2.033,0.551-1.152-0.402  c0.649-0.678,1.425-1.907,1.425-2.267c0-0.063-0.314,0.042-0.671,0.233c-0.377,0.212-1.215,0.53-1.844,0.72l-1.131,0.361l-1.027-0.7  c-0.566-0.381-1.361-0.805-1.781-0.932C39.766,21.902,38.131,21.944,37.167,22.283z M33,64C16.432,64,3,50.569,3,34S16.432,4,33,4  s30,13.431,30,30S49.568,64,33,64z" style="fill-rule:evenodd;clip-rule:evenodd;"/></svg>
</span>
        
      </a>
    
  </div>


      <h1 class="f1 athelas mt3 mb1">Data Wrangling with R Basics - 2</h1>
      
      <p class="tracked">
        By <strong>Yaoyu E. Wang</strong>
      </p>
      
      
      
      <time class="f6 mv4 dib tracked" datetime="2022-04-25T00:00:00Z">April 25, 2022</time>
      

      
      
    </header>
    <div class="nested-copy-line-height lh-copy serif f4 nested-links mid-gray pr4-l w-two-thirds-l"><h2 id="working-with-genomic-data-with-r">Working with genomic data with R</h2>
<p>We provided 5 files in the <strong>Data</strong> directory: -
differntial_results.csv - gene_expression.tsv - gene_annotations.tsv</p>
<ul>
<li>my_pathway_genes.txt - mutations.tsv</li>
</ul>
<p>These files are stored in either comma-delimited (csv) or tab-delimited
(tsv) formats. Files can be read with the following functions available
from default R libraries. These are built-in “functions” from R. All
read files, but with slightly different default behavior.</p>
<!-- raw HTML omitted -->
<pre><code># setwd(&quot;~/modules/data_wrangling&quot;)
# First, load all the data:
print('Loading data...')

## [1] &quot;Loading data...&quot;

dge_results &lt;- read.table('scripts and data/data/differential_results.csv', sep=',', header=T)
expressions = read.table('scripts and data/data/gene_expression.tsv', header=T)
annotations = read.delim('scripts and data/data/gene_annotations.tsv')
pathways = read.table('scripts and data/data/my_pathway_genes.txt', sep='\t', col.names=c('gene_name','pathway'))
mutations = read.delim('scripts and data/data/mutations.tsv')
print('Done loading data.')

## [1] &quot;Done loading data.&quot;
</code></pre>
<p>Reading files with read.table (or equivalents) automatically creates a
dataframe. R’s version of a Excel spreadsheet- contains different data
(of different types!) in the columns and “observations” in the rows.</p>
<h2 id="data-inspection">Data Inspection</h2>
<p>Data can be inspected after loaded with these common commands:<br>
<strong>class()</strong>: data type (e.g. character, numeric, etc.) of vectors and
data structure of dataframes, matrices, and lists. <strong>summary()</strong>:
detailed display, including descriptive statistics, frequencies
<strong>head()</strong>: will display the top 5 entries for the variable <strong>tail()</strong>:
will display the bottom 5 entries for the variable</p>
<p>Specifically, for Dataframe and matrix data structure: <strong>dim()</strong>:
returns dimensions of the dataset <strong>nrow()</strong>: returns the number of rows
in the dataset <strong>ncol()</strong>: returns the number of columns in the dataset
<strong>rownames()</strong>: returns the row names in the dataset <strong>colnames()</strong>:
returns the column names in the dataset</p>
<pre><code>print(&quot;Show the first six rows of each variable&quot;)

## [1] &quot;Show the first six rows of each variable&quot;

head(dge_results)

##         gene   baseMean        C          T foldChange log2FoldChange
## 1    DDX11L1  0.0000000  0.00000  0.0000000         NA             NA
## 2     WASH7P 35.6756001 36.87866 34.4725431  0.9347559    -0.09733839
## 3 MIR1302-10  0.0000000  0.00000  0.0000000         NA             NA
## 4    FAM138A  0.0000000  0.00000  0.0000000         NA             NA
## 5     OR4G4P  0.0000000  0.00000  0.0000000         NA             NA
## 6    OR4G11P  0.4759207  0.00000  0.9518413        Inf            Inf
##        pval padj
## 1        NA   NA
## 2 0.8764378    1
## 3        NA   NA
## 4        NA   NA
## 5        NA   NA
## 6 0.6676182    1

head(expressions)

##         gene SW1_Control SW2_Control SW3_Control SW4_Treated SW5_Treated
## 1    DDX11L1     0.00000     0.00000     0.00000     0.00000    0.000000
## 2     WASH7P    45.77888    22.84061    42.01649    50.16839   29.507082
## 3 MIR1302-10     0.00000     0.00000     0.00000     0.00000    0.000000
## 4    FAM138A     0.00000     0.00000     0.00000     0.00000    0.000000
## 5     OR4G4P     0.00000     0.00000     0.00000     0.00000    0.000000
## 6    OR4G11P     0.00000     0.00000     0.00000     0.00000    2.855524
##   SW6_Treated
## 1     0.00000
## 2    23.74216
## 3     0.00000
## 4     0.00000
## 5     0.00000
## 6     0.00000

head(annotations)

##   chrom start   end strand         name
## 1  chr1 11869 14409      +      DDX11L1
## 2  chr1 14404 29570      -       WASH7P
## 3  chr1 17369 17436      -    MIR6859-3
## 4  chr1 29554 31109      + RP11-34P13.3
## 5  chr1 30366 30503      +    MIR1302-9
## 6  chr1 34554 36081      -      FAM138A

head(pathways)

##   gene_name    pathway
## 1      PCK2 glycolysis
## 2      PCK1 glycolysis
## 3      FBP2 glycolysis
## 4      BPGM glycolysis
## 5   ALDH3A2 glycolysis
## 6   ALDH3A1 glycolysis

head(mutations)

##   chrom      pos ref alt
## 1     5 20578198   C   T
## 2     2  4642922   T   A
## 3     1 15947000   C   A
## 4    16  3573172   G   C
## 5    13 14306989   C   T
## 6    11  5028652   C   A

print(&quot;Done showing the first row of each data&quot;)

## [1] &quot;Done showing the first row of each data&quot;
</code></pre>
<h2 id="slicing-data">Slicing data</h2>
<p>To refer to a cell in Excel, you’re used to A1 (first row, first
column), C2 (third row, second column). Similar with a DataFrame:</p>
<pre><code>head(expressions)

##         gene SW1_Control SW2_Control SW3_Control SW4_Treated SW5_Treated
## 1    DDX11L1     0.00000     0.00000     0.00000     0.00000    0.000000
## 2     WASH7P    45.77888    22.84061    42.01649    50.16839   29.507082
## 3 MIR1302-10     0.00000     0.00000     0.00000     0.00000    0.000000
## 4    FAM138A     0.00000     0.00000     0.00000     0.00000    0.000000
## 5     OR4G4P     0.00000     0.00000     0.00000     0.00000    0.000000
## 6    OR4G11P     0.00000     0.00000     0.00000     0.00000    2.855524
##   SW6_Treated
## 1     0.00000
## 2    23.74216
## 3     0.00000
## 4     0.00000
## 5     0.00000
## 6     0.00000

# These two lines return the same number
expressions[2,2]

## [1] 45.77888

expressions[2, 'SW1_Control']

## [1] 45.77888

# Select First Column:
head(expressions[1])

##         gene
## 1    DDX11L1
## 2     WASH7P
## 3 MIR1302-10
## 4    FAM138A
## 5     OR4G4P
## 6    OR4G11P

head(expressions['gene'])  

##         gene
## 1    DDX11L1
## 2     WASH7P
## 3 MIR1302-10
## 4    FAM138A
## 5     OR4G4P
## 6    OR4G11P

head(expressions[,1])

## [1] DDX11L1    WASH7P     MIR1302-10 FAM138A    OR4G4P     OR4G11P   
## 56638 Levels: 5S_rRNA 7SK A1BG A1BG-AS1 A1CF A2M A2M-AS1 A2ML1 ... ZZZ3

head(expressions$gene)

## [1] DDX11L1    WASH7P     MIR1302-10 FAM138A    OR4G4P     OR4G11P   
## 56638 Levels: 5S_rRNA 7SK A1BG A1BG-AS1 A1CF A2M A2M-AS1 A2ML1 ... ZZZ3

# select First Row:
head(expressions[1,])

##      gene SW1_Control SW2_Control SW3_Control SW4_Treated SW5_Treated
## 1 DDX11L1           0           0           0           0           0
##   SW6_Treated
## 1           0

# select first three columns
head(expressions[1:3])

##         gene SW1_Control SW2_Control
## 1    DDX11L1     0.00000     0.00000
## 2     WASH7P    45.77888    22.84061
## 3 MIR1302-10     0.00000     0.00000
## 4    FAM138A     0.00000     0.00000
## 5     OR4G4P     0.00000     0.00000
## 6    OR4G11P     0.00000     0.00000

# select all columns except first:
head(expressions[2:ncol(expressions)])

##   SW1_Control SW2_Control SW3_Control SW4_Treated SW5_Treated SW6_Treated
## 1     0.00000     0.00000     0.00000     0.00000    0.000000     0.00000
## 2    45.77888    22.84061    42.01649    50.16839   29.507082    23.74216
## 3     0.00000     0.00000     0.00000     0.00000    0.000000     0.00000
## 4     0.00000     0.00000     0.00000     0.00000    0.000000     0.00000
## 5     0.00000     0.00000     0.00000     0.00000    0.000000     0.00000
## 6     0.00000     0.00000     0.00000     0.00000    2.855524     0.00000
</code></pre>
<h2 id="using-boolean">Using boolean</h2>
<ol>
<li>Let’s try a little example to select genes located on chromosome 7
and 3 from small gene annotation file ‘demo_annotations.tsv’. This
can be done in two ways:</li>
</ol>
<h4 id="excel">Excel</h4>
<pre><code>Open “demo_annotations.tsv” in excel
Insert table
Select ‘chrom’ column
Set ‘chrom’ column equal ‘chr7’
Set ‘chrom’ column equal ‘chr3’
</code></pre>
<h4 id="r">R</h4>
<p>In R using boolean comparison to create logical vector</p>
<pre><code>df=read.table(&quot;scripts and data/data/demo_annotations.tsv&quot;, header=T)
chroms=df$chrom
is_chr7 = chroms == 'chr7'
is_chr7

## [1] FALSE  TRUE FALSE  TRUE

df[is_chr7,]

##   chrom start end strand name
## 2  chr7   200 275      - TP53
## 4  chr7   500 600      + CD44

is_chr3 &lt;- chroms == 'chr3'
df[is_chr3,]

## [1] chrom  start  end    strand name  
## &lt;0 rows&gt; (or 0-length row.names)
</code></pre>
<ol>
<li>Select genes from gene annotations that are in your oncogene list
(‘KRAS’, ‘TP53’)</li>
</ol>
<h4 id="excel-1">Excel</h4>
<pre><code>#select gene annotations for (‘KRAS’, ‘TP53’)
Go to ‘name’ column
Find ‘KRAS’
Click on ‘KRAS’
Find ‘TP53’
Click on ‘TP53’
</code></pre>
<h4 id="r-1">R</h4>
<pre><code># select gene that are annotated as oncogenes use %in%
oncogenes = c('KRAS', 'TP53')  
is_oncogene = df$name %in% oncogenes
df[is_oncogene,]

##   chrom start end strand name
## 1  chr1   100 150      + KRAS
## 2  chr7   200 275      - TP53
</code></pre>
<ol>
<li>Select genes from gene annotations that are in your oncogene list
(‘KRAS’, ‘TP53’) and on chrom 7</li>
</ol>
<h4 id="excel-2">Excel</h4>
<pre><code>Go to ‘name’ column
Find ‘KRAS’
Click on ‘KRAS’
Find ‘TP53’
Click on ‘TP53’
Set ‘chrom’ column equal ‘chr7’
</code></pre>
<h4 id="r-2">R</h4>
<pre><code># Combine multiple criteria with AND (&amp;) operation
selection_criteria = is_oncogene &amp; is_chr7
data_subset = df[selection_criteria, ]
data_subset

##   chrom start end strand name
## 2  chr7   200 275      - TP53
</code></pre>
<p>End of toy data set…</p>
<h2 id="real-size-data-set">Real Size Data Set</h2>
<h4 id="1-use-the-data-set-loaded-earlier-in-the-script-to-identify-ras-signaling-genes-that-are-significantly-up-regulated-from-our-result-files">1) Use the data set loaded earlier in the script to identify ras signaling genes that are significantly up-regulated from our result files:</h4>
<p>Steps to find these genes:</p>
<ul>
<li>Get all the gene names that are in the pathway of interest
(ras_signaling) from ‘my_pathway_genes.tsv’</li>
<li>Find differential gene expression(DGE) results for those RAS genes
from ‘differential_results.csv’</li>
<li>Find all the differentially expressed genes an adjusted p &lt; 0.05
threshold from ‘differential_results.csv’</li>
<li>Find all the differentially expressed genes that are up-regulated
with log fold change (log2FoldChange) &gt; 0 from
‘differential_results.csv’</li>
<li>Keep only genes that pass all 3 “tests” (are True for all 3
conditions)</li>
<li>Just in case, remove missing data</li>
</ul>
<!-- raw HTML omitted -->
<pre><code>print('Filtering for ras signaling genes that are significantly upregulated.')

## [1] &quot;Filtering for ras signaling genes that are significantly upregulated.&quot;

# Get a vector of gene names that are in the pathway of interest
ras_genes = pathways[pathways['pathway'] == 'ras_signaling', 'gene_name']

#Create a boolean (True or False) vector for those RAS genes
is_ras_gene = dge_results$gene %in% ras_genes

# Create a boolean vector for whether the gene is significantly changed
# at a p &lt; 0.05 threshold
is_significant = dge_results$padj &lt; 0.05

# Create a boolean vector for whether the gene is upregulated
is_upregulated = dge_results$log2FoldChange &gt; 0

# Keep only rows that pass all 3 &quot;tests&quot; (are True for all 3)
selected_rows = dge_results[is_ras_gene &amp; is_significant &amp; is_upregulated, ]

# Just in case, remove missing data
selected_rows = na.omit(selected_rows)

# We now have the rows/genes-- now, keep only a subset of the columns:
upregulated_ras_genes = selected_rows[c('gene','baseMean','log2FoldChange','padj')]
print('Done filtering genes.')

## [1] &quot;Done filtering genes.&quot;
</code></pre>
<h4 id="2-find-ras-signaling-genes-that-are-significantly-up-regulated">2) Find ras signaling genes that are significantly up-regulated</h4>
<p>with gene coordinates and gene expression values from our result files</p>
<ul>
<li>Merge result with ‘gene_annotations.tsv’ on the selected genes</li>
<li>Merge result with ‘gene_expression.tsv’ on the selected genes</li>
</ul>
<!-- raw HTML omitted -->
<pre><code># Merge with gene coordinates
ras_up_genes_w_coords = merge(upregulated_ras_genes, annotations, by.x = 'gene', by.y='name')
# Merge with gene expression
ras_up_genes_w_coords_expression = merge(ras_up_genes_w_coords, expressions, by='gene')

# examine the variable
head(ras_up_genes_w_coords_expression)

##     gene   baseMean log2FoldChange         padj chrom     start       end
## 1 ANGPT1  1111.0617      0.5136897 2.783113e-04  chr8 107249482 107498055
## 2 BCL2L1 10000.0591      0.3472426 8.742192e-08 chr20  31664452  31723989
## 3  CALM2 73953.5081      0.2790400 5.369188e-06  chr2  47160082  47176601
## 4   FGF1   955.6761      0.2797321 2.553878e-02  chr5 142592178 142698070
## 5  FGF14   419.1095      0.9456159 7.832240e-05 chr13 101710804 102402457
## 6   FGF2 13101.6427      0.9214510 1.696858e-21  chr4 122826708 122898236
##   strand SW1_Control SW2_Control SW3_Control SW4_Treated SW5_Treated
## 1      -    765.5787    993.5663    986.8196   1237.1883   1274.5156
## 2      -   8944.4139   8877.7279   8584.9902  11402.5583  10930.9460
## 3      -  62166.7448  70176.7209  68128.0296  78758.1010  81115.9194
## 4      -    819.1497    879.3633    891.4309   1094.7459   1041.3144
## 5      -    216.2322    297.9661    345.2165    510.6425    502.5722
## 6      +   7597.3460   9885.8292   9679.6900  16386.2503  17452.9629
##   SW6_Treated
## 1   1408.7014
## 2  11259.7183
## 3  83375.5330
## 4   1008.0524
## 5    642.0275
## 6  17607.7777
</code></pre>
<h2 id="integrating-the-snp-data-with-expression-data">Integrating the SNP data with expression data</h2>
<p><strong>Problem</strong>: The chromosome notations are different and can not be
merged. chrom has chromosome as ‘chr8’ but the <em>mutations</em> has chrom as
‘8’.</p>
<p>Strings can be concatenated by <em>paste()</em> or <em>poste0()</em> | Function |
package |Description |—|—|—| |paste()|base|Concatenate vectors after
converting to character with &quot; &quot; as separator |paste0()|base| paste with
no separator, i.e. paste0(‘a’,‘b’) becomes ‘ab’</p>
<pre><code># Select the first row from mutation and change chrom from '5' to 'chr5'
i=1
chrom =mutations[i,'chrom']
chrom_w_prefix = paste('chr', chrom, sep='')  
mutations[i,'chrom'] =chrom_w_prefix
</code></pre>
<p>We can change one instance. but then do we want to change each row
one-by-one for 7479 times? There is a better way to do this…</p>
<h3 id="use-flow-control---the-for-loop">Use flow control - the “for loop”</h3>
<pre><code>print('Example using the for loop flow control')

## [1] &quot;Example using the for loop flow control&quot;

my_vector &lt;- c(10,11,12,13) 
for (item in my_vector){
  print(item)
  #do actual operations here
}

## [1] 10
## [1] 11
## [1] 12
## [1] 13
</code></pre>
<p>Everything between { and } is repeated. The script within the { } is
called a “block” of code. The indentation within the block is not
necessary, but helps with reading. The loop variable <em>item</em> is
arbitrary.</p>
<h3 id="changing-chromosome-names-in-mutation-dataframe">Changing chromosome names in mutation dataframe</h3>
<pre><code># We need to add the 'chr' prefix to the chromosome names that are in the mutations dataframe.
# This way, all the chromosome names are consistent.  
# The method below is a slow way to do this, but easier to understand.
for ( i in 2:nrow(mutations) )
{
    chrom = mutations[i,'chrom']
    # Add ‘chr’ to each chrom value
    chrom_w_prefix = paste('chr', chrom, sep='')
    mutations[i, 'chrom'] = chrom_w_prefix
}
</code></pre>
<p>Filter to keep only genes that are mutated. The idea is For each row (a
gene), see if any of the mutations are “inside” that gene e.g. Here, the
first mutation is on the KRAS gene</p>
<pre><code># Again, this a slow, but clear way to do this.
ras_up_mutated_genes = data.frame()                         # define an empty result variable
for ( i in 1:nrow(ras_up_genes_w_coords) )
{
  gene_info = ras_up_genes_w_coords[i,]               # get row i and save as gene_info
  same_chrom = gene_info$chrom == mutations$chrom       # get mutation row with same chrom
  past_start = mutations$pos &gt;gene_info$start             # get position &gt; then gene_info start
  before_end = mutations$pos &lt;gene_info$end             # get position &lt; then gene_info start   
  overlap = same_chrom &amp; past_start &amp; before_end        # get mutations fit all 3 conditions
  if ( any(overlap) )                                             # if there is any mutation quality
  {
    ras_up_mutated_genes = rbind(ras_up_mutated_genes, gene_info) # use rbind() to add to result variable
  }
}
</code></pre>
<h3 id="writing-files">Writing files</h3>
<p>Now, we want to save our clean results to a file. Look at the help from
R’s built-in write.table function.</p>
<!-- raw HTML omitted -->
<pre><code>print('Merge to add in the expression data')

## [1] &quot;Merge to add in the expression data&quot;

# note no by.x or by.y since the 'gene' column is in each dataframe
final_data = merge(ras_up_mutated_genes, expressions)
head(final_data)

##     gene    baseMean log2FoldChange         padj chrom     start       end
## 1  CALM2 73953.50812      0.2790400 5.369188e-06  chr2  47160082  47176601
## 2 MAPK10    99.11630      0.8205749 1.241397e-02  chr4  86015123  86594131
## 3  RAB5A  4408.97751      0.3199285 2.044410e-05  chr3  19947079  19985175
## 4   RGL1  3355.69859      0.2658060 1.414847e-03  chr1 183636085 183928531
## 5   RRAS  2591.26593      0.3950969 3.138126e-03 chr19  49635292  49640201
## 6    TEK    33.28648      1.4060046 1.770961e-02  chr9  27109141  27230175
##   strand SW1_Control SW2_Control SW3_Control SW4_Treated SW5_Treated
## 1      - 62166.74477 70176.72087 68128.02960 78758.10104 81115.91944
## 2      -    70.12935    77.86570    66.99926   145.12998   122.78753
## 3      +  3531.79188  4172.56326  4061.97223  4860.06277  4877.23505
## 4      +  3176.27504  3054.41182  2911.62896  3743.81609  3554.17558
## 5      -  2598.68194  2074.34222  2042.90971  3009.20753  2851.71667
## 6      +    22.40243    18.68777    13.62697    51.06425    51.39943
##   SW6_Treated
## 1 83375.53303
## 2   111.78599
## 3  4950.23988
## 4  3693.88403
## 5  2970.73748
## 6    42.53803
</code></pre>
<p>We will create an ‘Outputs’ directory, if not exist, to store all of the
output files.</p>
<pre><code># get current working directory. 
getwd()

## [1] &quot;/Users/yaoyuwang/Dropbox (Harvard University)/qBRC-Share/Presentations/QBRC-Workshops/Data Wrangling with R/2022-0424-IID&quot;

# Current directory is Data directory, change to output
if(!dir.exists('Outputs')){
  dir.create('Outputs')
}
# write final data out as a tab-seperated-value file (.tsv)
print('Write the final data to file.')

## [1] &quot;Write the final data to file.&quot;

write.table(final_data, 'Outputs/final_data.tsv', sep='\t', quote=F)
</code></pre>
<p>write out file in Excel format using <strong>writexl</strong>. Check if <strong>writexl</strong>
package is available, if not install the package using function
<em>install.packages()</em>.</p>
<pre><code>if(!require(readxl)) install.packages(&quot;readxl&quot;)
if(!require(writexl)) install.packages(&quot;writexl&quot;)

library(readxl)    # load  library to write excel file
library(writexl)    # load  library to write excel file

print('Write the final data to Excel file.')
# start a new excel file and write final results 
write_xlsx(final_data, &quot;Outputs/final_results.xlsx&quot;, col_names=TRUE)
# add a new sheet with new data onto the file we just created
write_xlsx(ras_up_genes_w_coords, &quot;Outputs/ras_up_genes_w_coords.xlsx&quot;, col_names=TRUE)
</code></pre>
<ul class="pa0">
  
   <li class="list di">
     <a href="/tags/r" class="link f5 grow no-underline br-pill ba ph3 pv2 mb2 dib black sans-serif">R</a>
   </li>
  
   <li class="list di">
     <a href="/tags/data-wrangling" class="link f5 grow no-underline br-pill ba ph3 pv2 mb2 dib black sans-serif">Data Wrangling</a>
   </li>
  
   <li class="list di">
     <a href="/tags/tutorial" class="link f5 grow no-underline br-pill ba ph3 pv2 mb2 dib black sans-serif">tutorial</a>
   </li>
  
</ul>
<div class="mt6 instapaper_ignoref">
      
      
      </div>
    </div>

    <aside class="w-30-l mt6-l">




  <div class="bg-light-gray pa3 nested-list-reset nested-copy-line-height nested-links">
    <p class="f5 b mb3">Related</p>
    <ul class="pa0 list">
	   
	     <li  class="mb2">
          <a href="/posts/datawranglingbasic-1/">Data Wrangling with R Basics - 1</a>
        </li>
	    
    </ul>
</div>

</aside>

  </article>

    </main>
    <footer class="bg-black bottom-0 w-100 pa3" role="contentinfo">
  <div class="flex justify-between">
  <a class="f4 fw4 hover-white no-underline white-70 dn dib-ns pv2 ph3" href="https://yaoyu-e-wang.github.io/" >
    &copy;  Ananke 2022 
  </a>
    <div>
<div class="ananke-socials">
  
    <a href="https://twitter.com/GoHugoIO" target="_blank" class="twitter ananke-social-link link-transition stackoverflow link dib z-999 pt3 pt0-l mr1" title="Twitter link" rel="noopener" aria-label="follow on Twitter——Opens in a new window">
      
        <span class="icon"><svg style="enable-background:new 0 0 67 67;" version="1.1" viewBox="0 0 67 67"  xml:space="preserve" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"><path d="M37.167,22.283c-2.619,0.953-4.274,3.411-4.086,6.101  l0.063,1.038l-1.048-0.127c-3.813-0.487-7.145-2.139-9.974-4.915l-1.383-1.377l-0.356,1.017c-0.754,2.267-0.272,4.661,1.299,6.271  c0.838,0.89,0.649,1.017-0.796,0.487c-0.503-0.169-0.943-0.296-0.985-0.233c-0.146,0.149,0.356,2.076,0.754,2.839  c0.545,1.06,1.655,2.097,2.871,2.712l1.027,0.487l-1.215,0.021c-1.173,0-1.215,0.021-1.089,0.467  c0.419,1.377,2.074,2.839,3.918,3.475l1.299,0.444l-1.131,0.678c-1.676,0.976-3.646,1.526-5.616,1.568  C19.775,43.256,19,43.341,19,43.405c0,0.211,2.557,1.397,4.044,1.864c4.463,1.377,9.765,0.783,13.746-1.568  c2.829-1.673,5.657-5,6.978-8.221c0.713-1.716,1.425-4.851,1.425-6.354c0-0.975,0.063-1.102,1.236-2.267  c0.692-0.678,1.341-1.419,1.467-1.631c0.21-0.403,0.188-0.403-0.88-0.043c-1.781,0.636-2.033,0.551-1.152-0.402  c0.649-0.678,1.425-1.907,1.425-2.267c0-0.063-0.314,0.042-0.671,0.233c-0.377,0.212-1.215,0.53-1.844,0.72l-1.131,0.361l-1.027-0.7  c-0.566-0.381-1.361-0.805-1.781-0.932C39.766,21.902,38.131,21.944,37.167,22.283z M33,64C16.432,64,3,50.569,3,34S16.432,4,33,4  s30,13.431,30,30S49.568,64,33,64z" style="fill-rule:evenodd;clip-rule:evenodd;"/></svg>
</span>
      
<span class="new-window"><svg  height="8px"  style="enable-background:new 0 0 1000 1000;" version="1.1" viewBox="0 0 1000 1000"  xml:space="preserve" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" >
<path d="M598 128h298v298h-86v-152l-418 418-60-60 418-418h-152v-86zM810 810v-298h86v298c0 46-40 86-86 86h-596c-48 0-86-40-86-86v-596c0-46 38-86 86-86h298v86h-298v596h596z" style="fill-rule:evenodd;clip-rule:evenodd;"/>
</svg>
</span></a>
  
</div>
</div>
  </div>
</footer>

  </body>
</html>
