<!DOCTYPE html>
<html lang="en-us">
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1">
    
    <title>Data Wrangling with R Basics - 3 | Ananke</title>
    <meta name="viewport" content="width=device-width,minimum-scale=1">
    <meta name="description" content="Data Wrangling with R Basics - 3 Let&rsquo;s load the dataset
data=list( dge_results=read.table(&#39;scripts and data/data/differential_results.csv&#39;, sep=&#39;,&#39;, header=T), expr=read.table(&#39;scripts and data/data/gene_expression.tsv&#39;, header=T, row.names=1), # first column being the rowname annot=read.delim(&#39;scripts and data/data/gene_annotations.tsv&#39;) ) find genes with greater than 0 mean expression between a given chr coordinate Let’s find out mean first You can use rowMeans, rowMax, rowMin to determine means for a data matrix
expr_mean=rowMeans(data$expr) gene_expr=cbind(data$expr, mean=expr_mean) We can use the function subset to find the given with a given coordinate">
    <meta name="generator" content="Hugo 0.101.0" />
    
    
    
    
      <meta name="robots" content="noindex, nofollow">
    

    
<link rel="stylesheet" href="/ananke/css/main.min.css" >



    
    
    
      

    

    
    
    <meta property="og:title" content="Data Wrangling with R Basics - 3" />
<meta property="og:description" content="Data Wrangling with R Basics - 3 Let&rsquo;s load the dataset
data=list( dge_results=read.table(&#39;scripts and data/data/differential_results.csv&#39;, sep=&#39;,&#39;, header=T), expr=read.table(&#39;scripts and data/data/gene_expression.tsv&#39;, header=T, row.names=1), # first column being the rowname annot=read.delim(&#39;scripts and data/data/gene_annotations.tsv&#39;) ) find genes with greater than 0 mean expression between a given chr coordinate Let’s find out mean first You can use rowMeans, rowMax, rowMin to determine means for a data matrix
expr_mean=rowMeans(data$expr) gene_expr=cbind(data$expr, mean=expr_mean) We can use the function subset to find the given with a given coordinate" />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://yaoyu-e-wang.github.io/posts/datawranglingbasic-3/" /><meta property="article:section" content="posts" />
<meta property="article:published_time" content="2022-04-26T00:00:00+00:00" />
<meta property="article:modified_time" content="2022-04-26T00:00:00+00:00" /><meta property="og:site_name" content="Ananke" />

<meta itemprop="name" content="Data Wrangling with R Basics - 3">
<meta itemprop="description" content="Data Wrangling with R Basics - 3 Let&rsquo;s load the dataset
data=list( dge_results=read.table(&#39;scripts and data/data/differential_results.csv&#39;, sep=&#39;,&#39;, header=T), expr=read.table(&#39;scripts and data/data/gene_expression.tsv&#39;, header=T, row.names=1), # first column being the rowname annot=read.delim(&#39;scripts and data/data/gene_annotations.tsv&#39;) ) find genes with greater than 0 mean expression between a given chr coordinate Let’s find out mean first You can use rowMeans, rowMax, rowMin to determine means for a data matrix
expr_mean=rowMeans(data$expr) gene_expr=cbind(data$expr, mean=expr_mean) We can use the function subset to find the given with a given coordinate"><meta itemprop="datePublished" content="2022-04-26T00:00:00+00:00" />
<meta itemprop="dateModified" content="2022-04-26T00:00:00+00:00" />
<meta itemprop="wordCount" content="674">
<meta itemprop="keywords" content="R,Data Wrangling,tutorial," /><meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="Data Wrangling with R Basics - 3"/>
<meta name="twitter:description" content="Data Wrangling with R Basics - 3 Let&rsquo;s load the dataset
data=list( dge_results=read.table(&#39;scripts and data/data/differential_results.csv&#39;, sep=&#39;,&#39;, header=T), expr=read.table(&#39;scripts and data/data/gene_expression.tsv&#39;, header=T, row.names=1), # first column being the rowname annot=read.delim(&#39;scripts and data/data/gene_annotations.tsv&#39;) ) find genes with greater than 0 mean expression between a given chr coordinate Let’s find out mean first You can use rowMeans, rowMax, rowMin to determine means for a data matrix
expr_mean=rowMeans(data$expr) gene_expr=cbind(data$expr, mean=expr_mean) We can use the function subset to find the given with a given coordinate"/>

	
  </head>

  <body class="ma0 avenir bg-near-white">

    
   
  

  <header>
    <div class="bg-black">
      <nav class="pv3 ph3 ph4-ns" role="navigation">
  <div class="flex-l justify-between items-center center">
    <a href="/" class="f3 fw2 hover-white no-underline white-90 dib">
      
        Ananke
      
    </a>
    <div class="flex-l items-center">
      

      
        <ul class="pl0 mr3">
          
          <li class="list f5 f4-ns fw4 dib pr3">
            <a class="hover-white no-underline white-90" href="/posts/" title="Posts page">
              Posts
            </a>
          </li>
          
        </ul>
      
      
<div class="ananke-socials">
  
    <a href="https://twitter.com/GoHugoIO" target="_blank" class="twitter ananke-social-link link-transition stackoverflow link dib z-999 pt3 pt0-l mr1" title="Twitter link" rel="noopener" aria-label="follow on Twitter——Opens in a new window">
      
        <span class="icon"><svg style="enable-background:new 0 0 67 67;" version="1.1" viewBox="0 0 67 67"  xml:space="preserve" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"><path d="M37.167,22.283c-2.619,0.953-4.274,3.411-4.086,6.101  l0.063,1.038l-1.048-0.127c-3.813-0.487-7.145-2.139-9.974-4.915l-1.383-1.377l-0.356,1.017c-0.754,2.267-0.272,4.661,1.299,6.271  c0.838,0.89,0.649,1.017-0.796,0.487c-0.503-0.169-0.943-0.296-0.985-0.233c-0.146,0.149,0.356,2.076,0.754,2.839  c0.545,1.06,1.655,2.097,2.871,2.712l1.027,0.487l-1.215,0.021c-1.173,0-1.215,0.021-1.089,0.467  c0.419,1.377,2.074,2.839,3.918,3.475l1.299,0.444l-1.131,0.678c-1.676,0.976-3.646,1.526-5.616,1.568  C19.775,43.256,19,43.341,19,43.405c0,0.211,2.557,1.397,4.044,1.864c4.463,1.377,9.765,0.783,13.746-1.568  c2.829-1.673,5.657-5,6.978-8.221c0.713-1.716,1.425-4.851,1.425-6.354c0-0.975,0.063-1.102,1.236-2.267  c0.692-0.678,1.341-1.419,1.467-1.631c0.21-0.403,0.188-0.403-0.88-0.043c-1.781,0.636-2.033,0.551-1.152-0.402  c0.649-0.678,1.425-1.907,1.425-2.267c0-0.063-0.314,0.042-0.671,0.233c-0.377,0.212-1.215,0.53-1.844,0.72l-1.131,0.361l-1.027-0.7  c-0.566-0.381-1.361-0.805-1.781-0.932C39.766,21.902,38.131,21.944,37.167,22.283z M33,64C16.432,64,3,50.569,3,34S16.432,4,33,4  s30,13.431,30,30S49.568,64,33,64z" style="fill-rule:evenodd;clip-rule:evenodd;"/></svg>
</span>
      
<span class="new-window"><svg  height="8px"  style="enable-background:new 0 0 1000 1000;" version="1.1" viewBox="0 0 1000 1000"  xml:space="preserve" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" >
<path d="M598 128h298v298h-86v-152l-418 418-60-60 418-418h-152v-86zM810 810v-298h86v298c0 46-40 86-86 86h-596c-48 0-86-40-86-86v-596c0-46 38-86 86-86h298v86h-298v596h596z" style="fill-rule:evenodd;clip-rule:evenodd;"/>
</svg>
</span></a>
  
</div>

    </div>
  </div>
</nav>

    </div>
  </header>



    <main class="pb7" role="main">
      
  
  <article class="flex-l flex-wrap justify-between mw8 center ph3">
    <header class="mt4 w-100">
      <aside class="instapaper_ignoref b helvetica tracked">
          
        POSTS
      </aside>
      










  <div id="sharing" class="mt3 ananke-socials">
    
      
      <a href="https://twitter.com/share?url=https://yaoyu-e-wang.github.io/posts/datawranglingbasic-3/&amp;text=Data%20Wrangling%20with%20R%20Basics%20-%203" class="ananke-social-link twitter no-underline" aria-label="share on Twitter">
        
        <span class="icon"> <svg style="enable-background:new 0 0 67 67;" version="1.1" viewBox="0 0 67 67"  xml:space="preserve" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"><path d="M37.167,22.283c-2.619,0.953-4.274,3.411-4.086,6.101  l0.063,1.038l-1.048-0.127c-3.813-0.487-7.145-2.139-9.974-4.915l-1.383-1.377l-0.356,1.017c-0.754,2.267-0.272,4.661,1.299,6.271  c0.838,0.89,0.649,1.017-0.796,0.487c-0.503-0.169-0.943-0.296-0.985-0.233c-0.146,0.149,0.356,2.076,0.754,2.839  c0.545,1.06,1.655,2.097,2.871,2.712l1.027,0.487l-1.215,0.021c-1.173,0-1.215,0.021-1.089,0.467  c0.419,1.377,2.074,2.839,3.918,3.475l1.299,0.444l-1.131,0.678c-1.676,0.976-3.646,1.526-5.616,1.568  C19.775,43.256,19,43.341,19,43.405c0,0.211,2.557,1.397,4.044,1.864c4.463,1.377,9.765,0.783,13.746-1.568  c2.829-1.673,5.657-5,6.978-8.221c0.713-1.716,1.425-4.851,1.425-6.354c0-0.975,0.063-1.102,1.236-2.267  c0.692-0.678,1.341-1.419,1.467-1.631c0.21-0.403,0.188-0.403-0.88-0.043c-1.781,0.636-2.033,0.551-1.152-0.402  c0.649-0.678,1.425-1.907,1.425-2.267c0-0.063-0.314,0.042-0.671,0.233c-0.377,0.212-1.215,0.53-1.844,0.72l-1.131,0.361l-1.027-0.7  c-0.566-0.381-1.361-0.805-1.781-0.932C39.766,21.902,38.131,21.944,37.167,22.283z M33,64C16.432,64,3,50.569,3,34S16.432,4,33,4  s30,13.431,30,30S49.568,64,33,64z" style="fill-rule:evenodd;clip-rule:evenodd;"/></svg>
</span>
        
      </a>
    
  </div>


      <h1 class="f1 athelas mt3 mb1">Data Wrangling with R Basics - 3</h1>
      
      <p class="tracked">
        By <strong>Yaoyu E. Wang</strong>
      </p>
      
      
      
      <time class="f6 mv4 dib tracked" datetime="2022-04-26T00:00:00Z">April 26, 2022</time>
      

      
      
    </header>
    <div class="nested-copy-line-height lh-copy serif f4 nested-links mid-gray pr4-l w-two-thirds-l"><h2 id="data-wrangling-with-r-basics---3">Data Wrangling with R Basics - 3</h2>
<p>Let&rsquo;s load the dataset</p>
<pre><code>data=list(
    dge_results=read.table('scripts and data/data/differential_results.csv', sep=',', header=T),
    expr=read.table('scripts and data/data/gene_expression.tsv', header=T, row.names=1),  # first column being the rowname
    annot=read.delim('scripts and data/data/gene_annotations.tsv')
)
</code></pre>
<h1 id="find-genes-with-greater-than-0-mean-expression-between-a-given-chr-coordinate">find genes with greater than 0 mean expression between a given chr coordinate</h1>
<h2 id="lets-find-out-mean-first">Let’s find out mean first</h2>
<p>You can use <strong>rowMeans</strong>, <strong>rowMax</strong>, <strong>rowMin</strong> to determine means for
a data matrix</p>
<pre><code>expr_mean=rowMeans(data$expr)
gene_expr=cbind(data$expr, mean=expr_mean)
</code></pre>
<p>We can use the function <strong>subset</strong> to find the given with a given
coordinate</p>
<pre><code># select all genes on chromosome 20
head(subset(data$annot, chrom=='chr20'))

##       chrom  start    end strand       name
## 53155 chr20  87250  97094      +    DEFB125
## 53156 chr20 142369 145751      +    DEFB126
## 53157 chr20 157470 159163      +    DEFB127
## 53158 chr20 173229 173296      - AL360078.1
## 53159 chr20 187853 189681      -    DEFB128
## 53160 chr20 227258 229886      +    DEFB129

# select all genes on chromosome 20 and on positive strand
head(subset(data$annot, chrom == 'chr20' &amp; strand==&quot;+&quot;))

##       chrom  start    end strand          name
## 53155 chr20  87250  97094      +       DEFB125
## 53156 chr20 142369 145751      +       DEFB126
## 53157 chr20 157470 159163      +       DEFB127
## 53160 chr20 227258 229886      +       DEFB129
## 53161 chr20 257736 261096      +       DEFB132
## 53162 chr20 267186 268857      + RP5-1103G7.10

# select all genes on chromosome 20 and on positive strand and between 10mb to 20mb
head(subset(data$annot, chrom == 'chr20' &amp; strand==&quot;+&quot; &amp; start&gt;10000000 &amp; end&lt;20000000))

##       chrom    start      end strand         name
## 53383 chr20 10025917 10026168      +  Metazoa_SRP
## 53384 chr20 10173520 10196990      + RP11-416N4.4
## 53385 chr20 10180235 10185775      + RP11-416N4.1
## 53386 chr20 10218830 10307418      +       SNAP25
## 53387 chr20 10334419 10334698      +    HIGD1AP15
## 53389 chr20 10385779 10387806      +      SDAD1P2

# select all genes on chromosome 20 and on positive strand and between 10mb to 20mb
head(subset(data$annot, chrom == 'chr20' &amp; strand==&quot;+&quot; &amp; start&gt;10000000 &amp; end&lt;20000000, select=c(&quot;chrom&quot;, &quot;name&quot;)))

##       chrom         name
## 53383 chr20  Metazoa_SRP
## 53384 chr20 RP11-416N4.4
## 53385 chr20 RP11-416N4.1
## 53386 chr20       SNAP25
## 53387 chr20    HIGD1AP15
## 53389 chr20      SDAD1P2

# store into a variable
chr20_subset=subset(data$annot, chrom == 'chr20' &amp; strand==&quot;+&quot; &amp; start&gt;10000000 &amp; end&lt;20000000, select=c(&quot;chrom&quot;, &quot;name&quot;))

# merge with gene expression data with means
chr20_expr=merge(chr20_subset, gene_expr, by.x=&quot;name&quot;, by.y=0)  #by.y=0 means using row names

head(chr20_expr)

##         name chrom SW1_Control SW2_Control SW3_Control SW4_Treated
## 1    AIMP1P1 chr20       0.000       0.000       0.000       0.000
## 2 AL035045.1 chr20       0.000       0.000       0.000       0.000
## 3 AL136090.1 chr20       0.000       0.000       0.000       0.000
## 4      BANF2 chr20       0.000       0.000       0.000       0.000
## 5      BTBD3 chr20    2309.398    2782.401    2790.122    2900.808
## 6  C20orf187 chr20       0.000       0.000       0.000       0.000
##   SW5_Treated  SW6_Treated         mean
## 1       0.000    0.9892566    0.1648761
## 2       0.000    0.0000000    0.0000000
## 3       0.000    0.0000000    0.0000000
## 4       0.000    0.0000000    0.0000000
## 5    3200.091 2960.8449159 2823.9441132
## 6       0.000    0.0000000    0.0000000
</code></pre>
<p>We can put these steps into <strong>function</strong> to make it more reusable:
<strong>function</strong> takes the form of: function( arglist ) expr return(value)</p>
<p>Let start with a simple function</p>
<pre><code>add&lt;-function(a,b){
  x=a+b
  return(x)
}

add(1,2)

## [1] 3

add(3,4)

## [1] 7

add(5,7)

## [1] 12
</code></pre>
<p>Let’s put everything together</p>
<pre><code>curr_coord=list(
  chr=&quot;chr20&quot;,
  strand=&quot;+&quot;,
  start=10000000,
  end=20000000
)

find_mean&lt;-function(data, gene_expr, curr_coord){
  selected_genes=subset(data$annot, chrom==curr_coord$chr &amp; 
                        strand==curr_coord$strand &amp; 
                          start&gt;curr_coord$start &amp; 
                          end&lt; curr_coord$end,
                         select=c(&quot;chrom&quot;, &quot;name&quot;))
  gene_mean_results=merge(selected_genes, gene_expr, by.x=&quot;name&quot;, by.y=0)
  return(gene_mean_results)
}

merge_results=find_mean(data, gene_expr, curr_coord)
head(merge_results==chr20_expr)

##      name chrom SW1_Control SW2_Control SW3_Control SW4_Treated
## [1,] TRUE  TRUE        TRUE        TRUE        TRUE        TRUE
## [2,] TRUE  TRUE        TRUE        TRUE        TRUE        TRUE
## [3,] TRUE  TRUE        TRUE        TRUE        TRUE        TRUE
## [4,] TRUE  TRUE        TRUE        TRUE        TRUE        TRUE
## [5,] TRUE  TRUE        TRUE        TRUE        TRUE        TRUE
## [6,] TRUE  TRUE        TRUE        TRUE        TRUE        TRUE
##      SW5_Treated SW6_Treated mean
## [1,]        TRUE        TRUE TRUE
## [2,]        TRUE        TRUE TRUE
## [3,]        TRUE        TRUE TRUE
## [4,]        TRUE        TRUE TRUE
## [5,]        TRUE        TRUE TRUE
## [6,]        TRUE        TRUE TRUE
</code></pre>
<p>Lastly, we filter out genes with means of 0</p>
<pre><code>final_results=subset(merge_results, mean&gt;0, select=c(&quot;name&quot;, &quot;chrom&quot;, &quot;mean&quot;))
head(final_results)

##       name chrom         mean
## 1  AIMP1P1 chr20 1.648761e-01
## 5    BTBD3 chr20 2.823944e+03
## 7  CSRP2BP chr20 1.096049e+03
## 8     DSTN chr20 3.369902e+04
## 9     DTD1 chr20 3.715940e+03
## 10 GAPDHP2 chr20 8.870973e+00

library(writexl)
write_xlsx(final_results, &quot;Outputs/final_results.xlsx&quot;)
</code></pre>
<ul class="pa0">
  
   <li class="list di">
     <a href="/tags/r" class="link f5 grow no-underline br-pill ba ph3 pv2 mb2 dib black sans-serif">R</a>
   </li>
  
   <li class="list di">
     <a href="/tags/data-wrangling" class="link f5 grow no-underline br-pill ba ph3 pv2 mb2 dib black sans-serif">Data Wrangling</a>
   </li>
  
   <li class="list di">
     <a href="/tags/tutorial" class="link f5 grow no-underline br-pill ba ph3 pv2 mb2 dib black sans-serif">tutorial</a>
   </li>
  
</ul>
<div class="mt6 instapaper_ignoref">
      
      
      </div>
    </div>

    <aside class="w-30-l mt6-l">




  <div class="bg-light-gray pa3 nested-list-reset nested-copy-line-height nested-links">
    <p class="f5 b mb3">Related</p>
    <ul class="pa0 list">
	   
	     <li  class="mb2">
          <a href="/posts/datawranglingbasic-2/">Data Wrangling with R Basics - 2</a>
        </li>
	    
	     <li  class="mb2">
          <a href="/posts/datawranglingbasic-1/">Data Wrangling with R Basics - 1</a>
        </li>
	    
    </ul>
</div>

</aside>

  </article>

    </main>
    <footer class="bg-black bottom-0 w-100 pa3" role="contentinfo">
  <div class="flex justify-between">
  <a class="f4 fw4 hover-white no-underline white-70 dn dib-ns pv2 ph3" href="https://yaoyu-e-wang.github.io/" >
    &copy;  Ananke 2022 
  </a>
    <div>
<div class="ananke-socials">
  
    <a href="https://twitter.com/GoHugoIO" target="_blank" class="twitter ananke-social-link link-transition stackoverflow link dib z-999 pt3 pt0-l mr1" title="Twitter link" rel="noopener" aria-label="follow on Twitter——Opens in a new window">
      
        <span class="icon"><svg style="enable-background:new 0 0 67 67;" version="1.1" viewBox="0 0 67 67"  xml:space="preserve" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"><path d="M37.167,22.283c-2.619,0.953-4.274,3.411-4.086,6.101  l0.063,1.038l-1.048-0.127c-3.813-0.487-7.145-2.139-9.974-4.915l-1.383-1.377l-0.356,1.017c-0.754,2.267-0.272,4.661,1.299,6.271  c0.838,0.89,0.649,1.017-0.796,0.487c-0.503-0.169-0.943-0.296-0.985-0.233c-0.146,0.149,0.356,2.076,0.754,2.839  c0.545,1.06,1.655,2.097,2.871,2.712l1.027,0.487l-1.215,0.021c-1.173,0-1.215,0.021-1.089,0.467  c0.419,1.377,2.074,2.839,3.918,3.475l1.299,0.444l-1.131,0.678c-1.676,0.976-3.646,1.526-5.616,1.568  C19.775,43.256,19,43.341,19,43.405c0,0.211,2.557,1.397,4.044,1.864c4.463,1.377,9.765,0.783,13.746-1.568  c2.829-1.673,5.657-5,6.978-8.221c0.713-1.716,1.425-4.851,1.425-6.354c0-0.975,0.063-1.102,1.236-2.267  c0.692-0.678,1.341-1.419,1.467-1.631c0.21-0.403,0.188-0.403-0.88-0.043c-1.781,0.636-2.033,0.551-1.152-0.402  c0.649-0.678,1.425-1.907,1.425-2.267c0-0.063-0.314,0.042-0.671,0.233c-0.377,0.212-1.215,0.53-1.844,0.72l-1.131,0.361l-1.027-0.7  c-0.566-0.381-1.361-0.805-1.781-0.932C39.766,21.902,38.131,21.944,37.167,22.283z M33,64C16.432,64,3,50.569,3,34S16.432,4,33,4  s30,13.431,30,30S49.568,64,33,64z" style="fill-rule:evenodd;clip-rule:evenodd;"/></svg>
</span>
      
<span class="new-window"><svg  height="8px"  style="enable-background:new 0 0 1000 1000;" version="1.1" viewBox="0 0 1000 1000"  xml:space="preserve" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" >
<path d="M598 128h298v298h-86v-152l-418 418-60-60 418-418h-152v-86zM810 810v-298h86v298c0 46-40 86-86 86h-596c-48 0-86-40-86-86v-596c0-46 38-86 86-86h298v86h-298v596h596z" style="fill-rule:evenodd;clip-rule:evenodd;"/>
</svg>
</span></a>
  
</div>
</div>
  </div>
</footer>

  </body>
</html>
