<!DOCTYPE html>
<html lang="en-us">
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1">
    
    <title>Basic RNASeq DGE Analysis | Ananke</title>
    <meta name="viewport" content="width=device-width,minimum-scale=1">
    <meta name="description" content="Begin RNA-Seq DGE Analysis We first load the RNA-Seq count matrix from ‘RNASeqData.Rdata’. This is a simple 3 vs 3 RNA-Seq experiment with Control vs Experiment in count matrix by gene.
# One can set up the path # Set working directory to PROJECT_DIR setwd(&quot;~/Dropbox (Harvard University)/qBRC-Share/Presentations/QBRC-Workshops/Intro to RNASeq Analysis/2022-0424-IID/RNASeq_Exploratory&quot;) # setwd(&quot;~/modules/rnaseq_dge/&quot;) load(&#39;RNASeq_exploratory.RData&#39;) ls() # list all current variables ## [1] &quot;count_data&quot; &quot;exploratory_results&quot; &quot;groups&quot; #head(count_data) # The head(count_data) provides the first 6 row of the &#39;count_data&#39; #head(groups) # The head(count_data) provides the first 6 row of the &#39;count_data&#39; Principle component analysis Once the matrix is loaded, we can perform principle component analysis to visualize sample distribution.">
    <meta name="generator" content="Hugo 0.101.0" />
    
    
    
    
      <meta name="robots" content="noindex, nofollow">
    

    
<link rel="stylesheet" href="/ananke/css/main.min.css" >



    
    
    
      

    

    
    
    <meta property="og:title" content="Basic RNASeq DGE Analysis" />
<meta property="og:description" content="Begin RNA-Seq DGE Analysis We first load the RNA-Seq count matrix from ‘RNASeqData.Rdata’. This is a simple 3 vs 3 RNA-Seq experiment with Control vs Experiment in count matrix by gene.
# One can set up the path # Set working directory to PROJECT_DIR setwd(&quot;~/Dropbox (Harvard University)/qBRC-Share/Presentations/QBRC-Workshops/Intro to RNASeq Analysis/2022-0424-IID/RNASeq_Exploratory&quot;) # setwd(&quot;~/modules/rnaseq_dge/&quot;) load(&#39;RNASeq_exploratory.RData&#39;) ls() # list all current variables ## [1] &quot;count_data&quot; &quot;exploratory_results&quot; &quot;groups&quot; #head(count_data) # The head(count_data) provides the first 6 row of the &#39;count_data&#39; #head(groups) # The head(count_data) provides the first 6 row of the &#39;count_data&#39; Principle component analysis Once the matrix is loaded, we can perform principle component analysis to visualize sample distribution." />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://yaoyu-e-wang.github.io/posts/rnaseq_dge/" /><meta property="article:section" content="posts" />
<meta property="article:published_time" content="2022-05-01T00:00:00+00:00" />
<meta property="article:modified_time" content="2022-05-01T00:00:00+00:00" /><meta property="og:site_name" content="Ananke" />

<meta itemprop="name" content="Basic RNASeq DGE Analysis">
<meta itemprop="description" content="Begin RNA-Seq DGE Analysis We first load the RNA-Seq count matrix from ‘RNASeqData.Rdata’. This is a simple 3 vs 3 RNA-Seq experiment with Control vs Experiment in count matrix by gene.
# One can set up the path # Set working directory to PROJECT_DIR setwd(&quot;~/Dropbox (Harvard University)/qBRC-Share/Presentations/QBRC-Workshops/Intro to RNASeq Analysis/2022-0424-IID/RNASeq_Exploratory&quot;) # setwd(&quot;~/modules/rnaseq_dge/&quot;) load(&#39;RNASeq_exploratory.RData&#39;) ls() # list all current variables ## [1] &quot;count_data&quot; &quot;exploratory_results&quot; &quot;groups&quot; #head(count_data) # The head(count_data) provides the first 6 row of the &#39;count_data&#39; #head(groups) # The head(count_data) provides the first 6 row of the &#39;count_data&#39; Principle component analysis Once the matrix is loaded, we can perform principle component analysis to visualize sample distribution."><meta itemprop="datePublished" content="2022-05-01T00:00:00+00:00" />
<meta itemprop="dateModified" content="2022-05-01T00:00:00+00:00" />
<meta itemprop="wordCount" content="1397">
<meta itemprop="keywords" content="R,RNASeq,DGE,tutorial," /><meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="Basic RNASeq DGE Analysis"/>
<meta name="twitter:description" content="Begin RNA-Seq DGE Analysis We first load the RNA-Seq count matrix from ‘RNASeqData.Rdata’. This is a simple 3 vs 3 RNA-Seq experiment with Control vs Experiment in count matrix by gene.
# One can set up the path # Set working directory to PROJECT_DIR setwd(&quot;~/Dropbox (Harvard University)/qBRC-Share/Presentations/QBRC-Workshops/Intro to RNASeq Analysis/2022-0424-IID/RNASeq_Exploratory&quot;) # setwd(&quot;~/modules/rnaseq_dge/&quot;) load(&#39;RNASeq_exploratory.RData&#39;) ls() # list all current variables ## [1] &quot;count_data&quot; &quot;exploratory_results&quot; &quot;groups&quot; #head(count_data) # The head(count_data) provides the first 6 row of the &#39;count_data&#39; #head(groups) # The head(count_data) provides the first 6 row of the &#39;count_data&#39; Principle component analysis Once the matrix is loaded, we can perform principle component analysis to visualize sample distribution."/>

	
  </head>

  <body class="ma0 avenir bg-near-white">

    
   
  

  <header>
    <div class="bg-black">
      <nav class="pv3 ph3 ph4-ns" role="navigation">
  <div class="flex-l justify-between items-center center">
    <a href="/" class="f3 fw2 hover-white no-underline white-90 dib">
      
        Ananke
      
    </a>
    <div class="flex-l items-center">
      

      
        <ul class="pl0 mr3">
          
          <li class="list f5 f4-ns fw4 dib pr3">
            <a class="hover-white no-underline white-90" href="/posts/" title="Posts page">
              Posts
            </a>
          </li>
          
        </ul>
      
      
<div class="ananke-socials">
  
    <a href="https://twitter.com/GoHugoIO" target="_blank" class="twitter ananke-social-link link-transition stackoverflow link dib z-999 pt3 pt0-l mr1" title="Twitter link" rel="noopener" aria-label="follow on Twitter——Opens in a new window">
      
        <span class="icon"><svg style="enable-background:new 0 0 67 67;" version="1.1" viewBox="0 0 67 67"  xml:space="preserve" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"><path d="M37.167,22.283c-2.619,0.953-4.274,3.411-4.086,6.101  l0.063,1.038l-1.048-0.127c-3.813-0.487-7.145-2.139-9.974-4.915l-1.383-1.377l-0.356,1.017c-0.754,2.267-0.272,4.661,1.299,6.271  c0.838,0.89,0.649,1.017-0.796,0.487c-0.503-0.169-0.943-0.296-0.985-0.233c-0.146,0.149,0.356,2.076,0.754,2.839  c0.545,1.06,1.655,2.097,2.871,2.712l1.027,0.487l-1.215,0.021c-1.173,0-1.215,0.021-1.089,0.467  c0.419,1.377,2.074,2.839,3.918,3.475l1.299,0.444l-1.131,0.678c-1.676,0.976-3.646,1.526-5.616,1.568  C19.775,43.256,19,43.341,19,43.405c0,0.211,2.557,1.397,4.044,1.864c4.463,1.377,9.765,0.783,13.746-1.568  c2.829-1.673,5.657-5,6.978-8.221c0.713-1.716,1.425-4.851,1.425-6.354c0-0.975,0.063-1.102,1.236-2.267  c0.692-0.678,1.341-1.419,1.467-1.631c0.21-0.403,0.188-0.403-0.88-0.043c-1.781,0.636-2.033,0.551-1.152-0.402  c0.649-0.678,1.425-1.907,1.425-2.267c0-0.063-0.314,0.042-0.671,0.233c-0.377,0.212-1.215,0.53-1.844,0.72l-1.131,0.361l-1.027-0.7  c-0.566-0.381-1.361-0.805-1.781-0.932C39.766,21.902,38.131,21.944,37.167,22.283z M33,64C16.432,64,3,50.569,3,34S16.432,4,33,4  s30,13.431,30,30S49.568,64,33,64z" style="fill-rule:evenodd;clip-rule:evenodd;"/></svg>
</span>
      
<span class="new-window"><svg  height="8px"  style="enable-background:new 0 0 1000 1000;" version="1.1" viewBox="0 0 1000 1000"  xml:space="preserve" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" >
<path d="M598 128h298v298h-86v-152l-418 418-60-60 418-418h-152v-86zM810 810v-298h86v298c0 46-40 86-86 86h-596c-48 0-86-40-86-86v-596c0-46 38-86 86-86h298v86h-298v596h596z" style="fill-rule:evenodd;clip-rule:evenodd;"/>
</svg>
</span></a>
  
</div>

    </div>
  </div>
</nav>

    </div>
  </header>



    <main class="pb7" role="main">
      
  
  <article class="flex-l flex-wrap justify-between mw8 center ph3">
    <header class="mt4 w-100">
      <aside class="instapaper_ignoref b helvetica tracked">
          
        POSTS
      </aside>
      










  <div id="sharing" class="mt3 ananke-socials">
    
      
      <a href="https://twitter.com/share?url=https://yaoyu-e-wang.github.io/posts/rnaseq_dge/&amp;text=Basic%20RNASeq%20DGE%20Analysis" class="ananke-social-link twitter no-underline" aria-label="share on Twitter">
        
        <span class="icon"> <svg style="enable-background:new 0 0 67 67;" version="1.1" viewBox="0 0 67 67"  xml:space="preserve" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"><path d="M37.167,22.283c-2.619,0.953-4.274,3.411-4.086,6.101  l0.063,1.038l-1.048-0.127c-3.813-0.487-7.145-2.139-9.974-4.915l-1.383-1.377l-0.356,1.017c-0.754,2.267-0.272,4.661,1.299,6.271  c0.838,0.89,0.649,1.017-0.796,0.487c-0.503-0.169-0.943-0.296-0.985-0.233c-0.146,0.149,0.356,2.076,0.754,2.839  c0.545,1.06,1.655,2.097,2.871,2.712l1.027,0.487l-1.215,0.021c-1.173,0-1.215,0.021-1.089,0.467  c0.419,1.377,2.074,2.839,3.918,3.475l1.299,0.444l-1.131,0.678c-1.676,0.976-3.646,1.526-5.616,1.568  C19.775,43.256,19,43.341,19,43.405c0,0.211,2.557,1.397,4.044,1.864c4.463,1.377,9.765,0.783,13.746-1.568  c2.829-1.673,5.657-5,6.978-8.221c0.713-1.716,1.425-4.851,1.425-6.354c0-0.975,0.063-1.102,1.236-2.267  c0.692-0.678,1.341-1.419,1.467-1.631c0.21-0.403,0.188-0.403-0.88-0.043c-1.781,0.636-2.033,0.551-1.152-0.402  c0.649-0.678,1.425-1.907,1.425-2.267c0-0.063-0.314,0.042-0.671,0.233c-0.377,0.212-1.215,0.53-1.844,0.72l-1.131,0.361l-1.027-0.7  c-0.566-0.381-1.361-0.805-1.781-0.932C39.766,21.902,38.131,21.944,37.167,22.283z M33,64C16.432,64,3,50.569,3,34S16.432,4,33,4  s30,13.431,30,30S49.568,64,33,64z" style="fill-rule:evenodd;clip-rule:evenodd;"/></svg>
</span>
        
      </a>
    
  </div>


      <h1 class="f1 athelas mt3 mb1">Basic RNASeq DGE Analysis</h1>
      
      <p class="tracked">
        By <strong>Yaoyu E. Wang</strong>
      </p>
      
      
      
      <time class="f6 mv4 dib tracked" datetime="2022-05-01T00:00:00Z">May 1, 2022</time>
      

      
      
    </header>
    <div class="nested-copy-line-height lh-copy serif f4 nested-links mid-gray pr4-l w-two-thirds-l"><h1 id="begin-rna-seq-dge-analysis">Begin RNA-Seq DGE Analysis</h1>
<p>We first load the RNA-Seq count matrix from ‘RNASeqData.Rdata’. This is
a simple 3 vs 3 RNA-Seq experiment with Control vs Experiment in count
matrix by gene.</p>
<pre><code># One can set up the path
# Set working directory to PROJECT_DIR
setwd(&quot;~/Dropbox (Harvard University)/qBRC-Share/Presentations/QBRC-Workshops/Intro to RNASeq Analysis/2022-0424-IID/RNASeq_Exploratory&quot;)
# setwd(&quot;~/modules/rnaseq_dge/&quot;)
load('RNASeq_exploratory.RData') 


ls()                       # list all current variables

## [1] &quot;count_data&quot;          &quot;exploratory_results&quot; &quot;groups&quot;

#head(count_data)           # The head(count_data) provides the first 6 row of the 'count_data'
#head(groups)           # The head(count_data) provides the first 6 row of the 'count_data'
</code></pre>
<h2 id="principle-component-analysis">Principle component analysis</h2>
<p>Once the matrix is loaded, we can perform principle component analysis
to visualize sample distribution.<br>
We first group samples into two vectors/variables: CONTROL and TREATED.</p>
<p>Calculate PCA use ‘prcomp’ command. Since prcomp compute PCA by the
rows, we will need to tranpose the count matrix such that the samples
are represented by rows and genes by the columns.</p>
<pre><code>CONTROL=groups$control$sample
TREATED=groups$treated$sample

pca=prcomp(t(count_data))        # compute PCA on transposed count_data
plot(pca$x[,1], pca$x[,2], xlab='pc1', ylab='pc2')
points(pca$x[CONTROL,1], pca$x[CONTROL,2],
       col=&quot;red&quot;, pch=18)
points(pca$x[TREATED,1], pca$x[TREATED,2],
       col=&quot;blue&quot;, pch=18)
</code></pre>
<p><img src="/images/RNASeq_DGE_files/figure-markdown_strict/unnamed-chunk-1-1.png" alt="PCA"></p>
<h2 id="hierarchical-clustering">Hierarchical Clustering</h2>
<p>We first calculate similarity matrix using Euclidean distance. The
matrix is computed by ‘dist’ function. ‘hclust’ function computes
hierarhical cluster tree and save it into an variable object ‘hctree’
that can be ploted by using generic ‘plot’ function.</p>
<pre><code># compute similarity matrix on the transposed count_data matrix using Euclidean distance
dist_matrix=dist(t(count_data), method=&quot;euclidean&quot;)
# compute hierarchical clustering tree and save it into variable called 'hctree'
hctree=hclust(dist_matrix)
# plot hctree
plot(hctree)
</code></pre>
<p><img src="/images/RNASeq_DGE_files/figure-markdown_strict/unnamed-chunk-2-1.png" alt=""></p>
<h2 id="differential-gene-expression">Differential Gene Expression</h2>
<p>We will use DESeq2 to perform differential gene expression. Since DESeq2
is a specialized bioconductor package/library, we will need to install
it before loading in the package to use.</p>
<p>The following command check if DESeq2 and Bioconductor Manager packages
have been installed, and install the packages if they have not been
installed.</p>
<pre><code># Check for Bioconductor, install it if it is not yet installed
  if (!requireNamespace('BiocManager', quietly = TRUE))
    install.packages('BiocManager')

# Check for DESeq2, install it if it is not yet installed
  if(!requireNamespace('DESeq2', quietly = TRUE))
    BiocManager::install('DESeq2', update=FALSE)
</code></pre>
<p>Once the packages are installed, we can call ‘library’ function to load
‘DESeq2’ library to use its functions.</p>
<p>We can then use the functions within DESeq2 to perform DGE analysis</p>
<p>First, we define which sampels are control group and which samples are
treated group The samples are ordered as YEW1_Control, YEW2_Control,
YEW3_Control, YEW1_Treated, YEW2_Treated, YEW3_Treated, so we can
use the following to define conditions of each sample</p>
<pre><code>condition &lt;- factor(c(&quot;control&quot;, &quot;control&quot;, &quot;control&quot;,
                      &quot;treated&quot;, &quot;treated&quot;, &quot;treated&quot;))
# alternatively, since the count_data is ordered in the same way as groups variable
# we can declare condition from groups:
# 
# condition &lt;- factor(c(as.character(groups$control$group), as.character(groups$treated$group)))
#
</code></pre>
<p>We then use the ‘condition’ to perform contrast on the count_data.
Noted that count_data is still raw count matrix, D ESeq2 performs
normalization and DGE together in one function.</p>
<p>If you want to know how to use a specific function in R, just type
‘?function’ in Console. For example type:</p>
<blockquote>
<p>?DESeqDataSetFromMatrix</p>
</blockquote>
<pre><code># Returns DGE gene list results in data frame format
dge_results=results(dds, contrast=c(&quot;condition&quot;, &quot;treated&quot;, &quot;control&quot;))

# make sure there is no missing values
dge_results=na.omit(dge_results)  

# filter for results with padj&lt;0.001 and absolute log2FoldChange&gt;2
filtered_results=subset(dge_results, padj&lt;0.001 &amp; abs(log2FoldChange)&gt;2)

# Obtain normalized count values
norm_count &lt;- counts(dds, normalized = T)

# output filtered_results
filtered_results

## log2 fold change (MLE): condition treated vs control 
## Wald test p-value: condition treated vs control 
## DataFrame with 30 rows and 6 columns
##                       baseMean    log2FoldChange             lfcSE
##                      &lt;numeric&gt;         &lt;numeric&gt;         &lt;numeric&gt;
## RP1-137D17.1  29.8839883584897  2.18080051514365 0.525441950834757
## AP000344.4      73.14903147842  2.12391047870749 0.333193245882354
## CTB-134H23.2  16.4631893142531 -4.06989298788851 0.902046395861882
## C12orf40      23.5935805512499 -2.69347617414478 0.622311943521575
## RP11-395P13.1 25.2555673433736  2.25522540567288  0.55251100155043
## ...                        ...               ...               ...
## RP11-413E6.7  62.9678012294187 -3.57669120749119 0.409461162253874
## CTD-2144E22.8 79.2703453044705 -2.48457438278124 0.339133969829025
## RP11-747D18.1  1062.4407047496  2.01064482797773  0.12042205283902
## RP11-864I4.4  25.0810724471142  2.31146493910142  0.55115609828326
## MSMP          44.7309465700437  2.16345289754786 0.400073627551603
##                            stat               pvalue                 padj
##                       &lt;numeric&gt;            &lt;numeric&gt;            &lt;numeric&gt;
## RP1-137D17.1   4.15041188028299 3.31877521851314e-05  0.00026025079086948
## AP000344.4      6.3744103608193 1.83668201976646e-10 3.60334193573427e-09
## CTB-134H23.2  -4.51184440906705 6.42663252483387e-06   5.821921826631e-05
## C12orf40      -4.32817689293055 1.50348695874195e-05  0.00012712159813363
## RP11-395P13.1  4.08177466031332 4.46931192753386e-05 0.000338444791804363
## ...                         ...                  ...                  ...
## RP11-413E6.7  -8.73511711783197 2.43403807314299e-18 1.20260893386853e-16
## CTD-2144E22.8 -7.32623270984574 2.36712611316642e-13 6.83284951698415e-12
## RP11-747D18.1  16.6966496632104 1.38633818987756e-62  7.2031449110832e-60
## RP11-864I4.4   4.19384807008606 2.74261710788168e-05 0.000220325854147882
## MSMP            5.4076368662136 6.38617383047045e-08 8.34139160148505e-07
</code></pre>
<p>We can compute PCA on normalized matrix again</p>
<pre><code>CONTROL=groups$control$sample
TREATED=groups$treated$sample

pca=prcomp(t(norm_count))        # compute PCA on transposed normalized_data
plot(pca$x[,1], pca$x[,2], xlab='pc1', ylab='pc2')
points(pca$x[CONTROL,1], pca$x[CONTROL,2],
       col=&quot;red&quot;, pch=18)
points(pca$x[TREATED,1], pca$x[TREATED,2],
       col=&quot;blue&quot;, pch=18)
</code></pre>
<p><img src="/images/RNASeq_DGE_files/figure-markdown_strict/norm_pca-1.png" alt=""></p>
<pre><code># compute similarity matrix on the transposed count_data matrix using Euclidean distance
dist_matrix=dist(t(norm_count), method=&quot;euclidean&quot;)
# compute hierarchical clustering tree and save it into variable called 'hctree'
hctree=hclust(dist_matrix)
# plot hctree
plot(hctree)
</code></pre>
<p><img src="/images/RNASeq_DGE_files/figure-markdown_strict/norm_pca-2.png" alt=""></p>
<h2 id="heatmap">Heatmap</h2>
<p>Generate heatmap using filtered dge results ‘filtered_results’. The
heatmap function is core function, but we want to have better coloring,
so we install and load gplots.</p>
<pre><code># Check and install &quot;gplots&quot; if not installed
if(!requireNamespace('gplots', quietly = TRUE))
  install.packages('gplots')
library(gplots)

## Warning: package 'gplots' was built under R version 3.5.2

## 
## Attaching package: 'gplots'

## The following object is masked from 'package:IRanges':
## 
##     space

## The following object is masked from 'package:S4Vectors':
## 
##     space

## The following object is masked from 'package:stats':
## 
##     lowess

DGE_matrix=norm_count[rownames(filtered_results),]
heatmap(DGE_matrix,col=colorpanel(50, 'blue', 'white', 'red'),
        Rowv=TRUE, Colv=TRUE)
</code></pre>
<p><img src="/images/RNASeq_DGE_files/figure-markdown_strict/heatmap-1.png" alt=""></p>
<pre><code># call the commmand again to write it into a file.
png(&quot;heatmap.png&quot;)
heatmap(DGE_matrix,col=colorpanel(50, 'blue', 'white', 'red'),
        Rowv=TRUE, Colv=TRUE)
dev.off()

## quartz_off_screen 
##                 2
</code></pre>
<h2 id="generate-volcano-plot">Generate Volcano Plot</h2>
<p>Again, we first check for the package ‘EnhancedVolcano’, and install it
if it is not already installed.</p>
<p>Load the library and run EnhancedVolcano function to generate volcano
plot.</p>
<pre><code>library(EnhancedVolcano)

## Warning: package 'EnhancedVolcano' was built under R version 3.5.2

## Loading required package: ggplot2

## Loading required package: ggrepel

## Warning: package 'ggrepel' was built under R version 3.5.2

volcano_plot&lt;-EnhancedVolcano(dge_results,
    lab = rownames(dge_results),
    x = 'log2FoldChange',
    y = 'padj',
    xlim = c(-5, 8),
    ylim = c(0, 20),
    #pCutoff = 10e-16,
    FCcutoff = 1.5
)

volcano_plot

## Warning: Removed 235 rows containing missing values (geom_point).

## Warning: Removed 11 rows containing missing values (geom_text).
</code></pre>
<p><img src="/images/RNASeq_DGE_files/figure-markdown_strict/volcano-1.png" alt=""></p>
<h2 id="gene-set-erichment-analysis">Gene Set Erichment Analysis</h2>
<p>There are many different packages retreiving MSigDB data to perform GSEA
analysis. Most of the packages perform these tasks in very similar ways
with different application programming interface (API). We will use the
following packages:</p>
<ul>
<li>migdbr
(<!-- raw HTML omitted --><a href="https://github.com/igordot/msigdbr">https://github.com/igordot/msigdbr</a><!-- raw HTML omitted -->)</li>
<li>fgsea
(<!-- raw HTML omitted --><a href="https://bioconductor.org/packages/release/bioc/html/fgsea.html">https://bioconductor.org/packages/release/bioc/html/fgsea.html</a><!-- raw HTML omitted -->)</li>
</ul>
<p>First, we install <strong>misgdbr</strong></p>
<pre><code># install misgdbr from CRAN
if(!require(msigdbr)) install.packages(&quot;msigdbr&quot;)

## Loading required package: msigdbr

library(msigdbr)
</code></pre>
<p>The <strong>misgdbr</strong> retrieve data from MSigDB database hosted at Broad
Institute
(<!-- raw HTML omitted --><a href="https://www.gsea-msigdb.org/gsea/msigdb/index.jsp">https://www.gsea-msigdb.org/gsea/msigdb/index.jsp</a><!-- raw HTML omitted -->).
The function <strong>misgdbr</strong> runs as the following:</p>
<p>####Usage msigdbr(species = “Homo sapiens”, category = NULL,
subcategory = NULL)</p>
<p>####Arguments |Parameters|Description| |:—:|—| |<strong>species</strong>|species
name, such as Homo sapiens, Mus musculus, etc.|
|<strong>category</strong>|collection, such as H, C1, C2, C3, C4, C5, C6, C7.|
|<strong>subcategory</strong>|sub-collection, such as CGP, MIR, BP, etc.|</p>
<pre><code>m_df = msigdbr(species = &quot;Homo sapiens&quot;, category = &quot;C2&quot;, subcategory = &quot;CP:KEGG&quot;)
# reformat the data frame into list for fgsea input
m_list = m_df %&gt;% split(x = .$gene_symbol, f = .$gs_name) 


# install fgsea
if(!require(fgsea)) BiocManager::install(&quot;fgsea&quot;)

## Loading required package: fgsea

## Loading required package: Rcpp

## Warning: package 'Rcpp' was built under R version 3.5.2

library(fgsea)

# reformat
gene_stat=dge_results$stat
names(gene_stat)=rownames(dge_results)
# sort genes
gene_stat=gene_stat[order(gene_stat)]

fgsea_results=fgsea(pathways=m_list, gene_stat, nperm=50, minSize=20)
fgsea_results=fgsea_results[order(fgsea_results$pval),]
head(fgsea_results)

##                                        pathway       pval padj         ES
## 1:    KEGG_ANTIGEN_PROCESSING_AND_PRESENTATION 0.03703704    1 -0.4975169
## 2:                          KEGG_AXON_GUIDANCE 0.03846154    1 -0.4746919
## 3:  KEGG_PHOSPHATIDYLINOSITOL_SIGNALING_SYSTEM 0.07692308    1 -0.4796340
## 4: KEGG_CYTOKINE_CYTOKINE_RECEPTOR_INTERACTION 0.11428571    1  0.3375730
## 5:                             KEGG_CELL_CYCLE 0.12903226    1  0.4658802
## 6:                 KEGG_SMALL_CELL_LUNG_CANCER 0.15384615    1 -0.4090299
##          NES nMoreExtreme size                                 leadingEdge
## 1: -1.487950            0   24      PDIA3,KLRC4,HLA-DOA,RFX5,KIR3DL1,IFNA4
## 2: -1.430389            0   25        CDK5,EPHA5,SEMA3E,PPP3R2,EFNA4,UNC5D
## 3: -1.503249            1   27 INPP5E,CALML6,CDS1,INPP5A,PLCE1,PIK3C2G,...
## 4:  1.271000            3   84  IL20,CCL28,TNFRSF9,ACVR2B,CXCL6,TNFSF4,...
## 5:  1.349746            3   37    CCNB3,SMAD4,RAD21,ZBTB17,MCM2,CDKN1C,...
## 6: -1.242841            3   26      MAX,TRAF2,COL4A1,LAMA5,PIAS1,ITGB1,...
</code></pre>
<p>Now write out the GSEA results</p>
<pre><code>library(writexl)    # load  library to write excel file

## Warning: package 'writexl' was built under R version 3.5.2

# Current directory is Data directory, change to output
if(!dir.exists('outputs')){
  dir.create('outputs')
}

# using ggplots's ggsave function to save the plot
print(&quot;Save the volcano plot in ggplot format&quot;)

## [1] &quot;Save the volcano plot in ggplot format&quot;

ggsave(&quot;outputs/volcano.png&quot;, plot=volcano_plot)

## Saving 7 x 5 in image

## Warning: Removed 235 rows containing missing values (geom_point).

## Warning: Removed 11 rows containing missing values (geom_text).

print('Write the final DGE data to Excel file.')

## [1] &quot;Write the final DGE data to Excel file.&quot;

# start a new excel file and write final results 
write_xlsx(as.data.frame(dge_results), &quot;outputs/dge_results.xlsx&quot;, col_names=TRUE)


print('Write the final GSEA data to Excel file.')

## [1] &quot;Write the final GSEA data to Excel file.&quot;

# start a new excel file and write final results 
write_xlsx(fgsea_results, &quot;outputs/gsea_results.xlsx&quot;, col_names=TRUE)
</code></pre>
<ul class="pa0">
  
   <li class="list di">
     <a href="/tags/r" class="link f5 grow no-underline br-pill ba ph3 pv2 mb2 dib black sans-serif">R</a>
   </li>
  
   <li class="list di">
     <a href="/tags/rnaseq" class="link f5 grow no-underline br-pill ba ph3 pv2 mb2 dib black sans-serif">RNASeq</a>
   </li>
  
   <li class="list di">
     <a href="/tags/dge" class="link f5 grow no-underline br-pill ba ph3 pv2 mb2 dib black sans-serif">DGE</a>
   </li>
  
   <li class="list di">
     <a href="/tags/tutorial" class="link f5 grow no-underline br-pill ba ph3 pv2 mb2 dib black sans-serif">tutorial</a>
   </li>
  
</ul>
<div class="mt6 instapaper_ignoref">
      
      
      </div>
    </div>

    <aside class="w-30-l mt6-l">




  <div class="bg-light-gray pa3 nested-list-reset nested-copy-line-height nested-links">
    <p class="f5 b mb3">Related</p>
    <ul class="pa0 list">
	   
	     <li  class="mb2">
          <a href="/posts/datawranglingbasic-3/">Data Wrangling with R Basics - 3</a>
        </li>
	    
	     <li  class="mb2">
          <a href="/posts/datawranglingbasic-2/">Data Wrangling with R Basics - 2</a>
        </li>
	    
	     <li  class="mb2">
          <a href="/posts/datawranglingbasic-1/">Data Wrangling with R Basics - 1</a>
        </li>
	    
    </ul>
</div>

</aside>

  </article>

    </main>
    <footer class="bg-black bottom-0 w-100 pa3" role="contentinfo">
  <div class="flex justify-between">
  <a class="f4 fw4 hover-white no-underline white-70 dn dib-ns pv2 ph3" href="https://yaoyu-e-wang.github.io/" >
    &copy;  Ananke 2022 
  </a>
    <div>
<div class="ananke-socials">
  
    <a href="https://twitter.com/GoHugoIO" target="_blank" class="twitter ananke-social-link link-transition stackoverflow link dib z-999 pt3 pt0-l mr1" title="Twitter link" rel="noopener" aria-label="follow on Twitter——Opens in a new window">
      
        <span class="icon"><svg style="enable-background:new 0 0 67 67;" version="1.1" viewBox="0 0 67 67"  xml:space="preserve" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"><path d="M37.167,22.283c-2.619,0.953-4.274,3.411-4.086,6.101  l0.063,1.038l-1.048-0.127c-3.813-0.487-7.145-2.139-9.974-4.915l-1.383-1.377l-0.356,1.017c-0.754,2.267-0.272,4.661,1.299,6.271  c0.838,0.89,0.649,1.017-0.796,0.487c-0.503-0.169-0.943-0.296-0.985-0.233c-0.146,0.149,0.356,2.076,0.754,2.839  c0.545,1.06,1.655,2.097,2.871,2.712l1.027,0.487l-1.215,0.021c-1.173,0-1.215,0.021-1.089,0.467  c0.419,1.377,2.074,2.839,3.918,3.475l1.299,0.444l-1.131,0.678c-1.676,0.976-3.646,1.526-5.616,1.568  C19.775,43.256,19,43.341,19,43.405c0,0.211,2.557,1.397,4.044,1.864c4.463,1.377,9.765,0.783,13.746-1.568  c2.829-1.673,5.657-5,6.978-8.221c0.713-1.716,1.425-4.851,1.425-6.354c0-0.975,0.063-1.102,1.236-2.267  c0.692-0.678,1.341-1.419,1.467-1.631c0.21-0.403,0.188-0.403-0.88-0.043c-1.781,0.636-2.033,0.551-1.152-0.402  c0.649-0.678,1.425-1.907,1.425-2.267c0-0.063-0.314,0.042-0.671,0.233c-0.377,0.212-1.215,0.53-1.844,0.72l-1.131,0.361l-1.027-0.7  c-0.566-0.381-1.361-0.805-1.781-0.932C39.766,21.902,38.131,21.944,37.167,22.283z M33,64C16.432,64,3,50.569,3,34S16.432,4,33,4  s30,13.431,30,30S49.568,64,33,64z" style="fill-rule:evenodd;clip-rule:evenodd;"/></svg>
</span>
      
<span class="new-window"><svg  height="8px"  style="enable-background:new 0 0 1000 1000;" version="1.1" viewBox="0 0 1000 1000"  xml:space="preserve" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" >
<path d="M598 128h298v298h-86v-152l-418 418-60-60 418-418h-152v-86zM810 810v-298h86v298c0 46-40 86-86 86h-596c-48 0-86-40-86-86v-596c0-46 38-86 86-86h298v86h-298v596h596z" style="fill-rule:evenodd;clip-rule:evenodd;"/>
</svg>
</span></a>
  
</div>
</div>
  </div>
</footer>

  </body>
</html>
